src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedArray`1.cs

```diff
@@ -0,0 +1,358 @@
+        }
+
+        bool IList.Contains(object value)
+        {
+            foreach (IList list in _items)
+            {
+                if (list.Contains(value))
+                    return true;
+            }
+
+            return false;
+        }
+
+        bool ICollection<T>.Contains(T value)
+        {
+            foreach (ICollection<T> collection in _items)
+            {
+                if (collection.Contains(value))
+                    return true;
+            }
+
+            return false;
+        }
+
+        int IList.IndexOf(object value)
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                IList list = _items[i];
+                var index = list.IndexOf(value);
+                if (index >= 0)
+                {
+                    return index + i * s_segmentSize;
+                }
+            }
+
+            return -1;
+        }
+
+        int IList<T>.IndexOf(T value)
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                IList<T> list = _items[i];
+                var index = list.IndexOf(value);
+                if (index >= 0)
+                {
+                    return index + i * s_segmentSize;
+                }
+            }
+
+            return -1;
+        }
+
+        void IList.Insert(int index, object value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void IList<T>.Insert(int index, T value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void IList.Remove(object value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        bool ICollection<T>.Remove(T value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void IList.RemoveAt(int index)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void IList<T>.RemoveAt(int index)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+            => GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator()
+            => GetEnumerator();
+
+        int IStructuralComparable.CompareTo(object other, IComparer comparer)
+        {
+            if (other is null)
+                return 1;
+
+            if (!(other is SegmentedArray<T> o)
+                || Length != o.Length)
+            {
+                throw new ArgumentException(CompilerExtensionsResources.ArgumentException_OtherNotArrayOfCorrectLength, nameof(other));
```

Comment:
why do the lengths need to be hte same?  i normally think of this is lexicographic fashion.  we should be able to campare differnt lengthed arrays in a sensible fashion.

Reply:
https://github.com/dotnet/runtime/blob/e0ec035994179e8ebd6ccf081711ee11d4c5491b/src/libraries/System.Private.CoreLib/src/System/Array.cs#L320-L323

Reply:
doc please